rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Delimiters.rsc

```diff
@@ -10,101 +10,162 @@ import util::Maybe;
+                return just(delimiter);
+            }
+            if (getOnlyFirst) {
+                return nothing();
             }
         }
-
-        index += (s: pairs); // Definitively added 
-        return pairs;
+        return nothing();
     }
 
-    return getDelimiterPairs(s);
-
-    // TODO: The current version of this function does not find delimiter pairs
-    // that are spread across multiple productions. For instance:
-    //
-    // ```
-    // lexical DelimitedNumber = Left Number Right;
-    //
-    // lexical Left   = "<";
-    // lexical Right  = ">";
-    // lexical Number = [0-9]+ !>> [0-9];
-    // ```
-    //
-    // In this example, `getDelimiterPairs(lex("Number"))` returns the empty
-    // set. This could be further improved.
+    solve (delimiters) delimiters = delimiters + (p: \do(p) | p <- todo());
+    return delimiters;
 }
 
+set[Production] getChildren(Grammar g, Symbol s)
+    = {child | child <- lookup(g, s)};
+
 @synopsis{
-    Gets the delimiter pair that encloses symbol `s` in a list, if any
+    Gets the unique rightmost delimiter (`begin`) and the unique leftmost
+    delimiter (`end`), if any, that occur **outside** production `p` in grammar
+    `g`.
+}
+
+@description{
+    For instance, consider the following grammar:
+
+    ```
+    lexical X  = Y;
+    lexical Y  = Y1 | Y2;
+    lexical Y1 = "[" Z "]"; 
+    lexical Y2 = "[" Z ")" [a-z];
+    lexical Z  = [a-z];
+    ```
+
+    The unique rightmost delimiter of the `Z` production is `[`. The remaining
+    productions do not have a unique rightmost delimiter.
+
+    The productions do not have a unique leftmost delimiter. In particular, the
+    `Z` productions has two leftmost delimiters, but not one unique.
+}
+
+DelimiterPair getOuterDelimiterPair(Grammar g, Production p)
+    = <getOuterDelimitersByProduction(g, backward())[p], getOuterDelimitersByProduction(g, forward())[p]>;
+
+@memo
+private map[Symbol, Maybe[Symbol]] getOuterDelimitersBySymbol(Grammar g, Direction direction) {
+    map[Symbol, Maybe[Symbol]] delimiters = (s: nothing() | /p: prod(_, _, _) := g, s := delabel(p.def));
+    set[Symbol] todo() = {s | s <- delimiters, delimiters[s] == nothing()};
+
+    Maybe[Symbol] \do(Symbol s)
+        = unique({\do(parent.def, rest) | parent <- getParents(g, s), [*_, /s, *rest] := reorder(parent.symbols, direction), /s !:= rest});
+
+    Maybe[Symbol] \do(Symbol def, list[Symbol] rest) {
+        for (s <- rest) {
+            s = delabel(s);
+            if (isDelimiter(s)) {
+                return just(s);
+            }
+            if (isNonTerminalType(s) && just(delimiter) := getInnerDelimitersBySymbol(g, direction)[s]) {
+                return just(delimiter);
+            }
+        }
+        return delimiters[delabel(def)];
+    }
+
+    solve (delimiters) delimiters = delimiters + (s: \do(s) | s <- todo());
```

Comment:
also here, I think the code will be a bit more readable if not written as 4 nested functions?